src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/AuditUtility.cs

```diff
@@ -0,0 +1,287 @@
+                            }
+
+                            if (!knownPackageVulnerabilities.TryGetValue(knownVulnerability, out List<string>? affectedGraphs))
+                            {
+                                affectedGraphs = new();
+                                knownPackageVulnerabilities.Add(knownVulnerability, affectedGraphs);
+                            }
+
+                            // Multiple package sources might list the same known vulnerability, so de-duple those too.
+                            if (!affectedGraphs.Contains(graph.TargetGraphName))
+                            {
+                                affectedGraphs.Add(graph.TargetGraphName);
+                            }
+                        }
+                    }
+                }
+            }
+            return result;
+        }
+
+        private async Task<GetVulnerabilityInfoResult?> GetAllVulnerabilityDataAsync(CancellationToken cancellationToken)
+        {
+            var results = new Task<GetVulnerabilityInfoResult?>[_vulnerabilityInfoProviders.Count];
+            for (int i = 0; i < _vulnerabilityInfoProviders.Count; i++)
+            {
+                IVulnerabilityInformationProvider provider = _vulnerabilityInfoProviders[i];
+                results[i] = provider.GetVulnerabilityInformationAsync(cancellationToken);
+            }
+
+            await Task.WhenAll(results);
+            if (cancellationToken.IsCancellationRequested)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            List<Exception>? errors = null;
+            List<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>>? knownVulnerabilities = null;
+            foreach (var resultTask in results)
+            {
+                var result = await resultTask;
+                if (result is null) continue;
+
+                if (result.KnownVulnerabilities != null)
+                {
+                    if (knownVulnerabilities == null)
+                    {
+                        knownVulnerabilities = new();
+                    }
+
+                    knownVulnerabilities.AddRange(result.KnownVulnerabilities);
+                }
+
+                if (result.Exceptions != null)
+                {
+                    if (errors == null)
+                    {
+                        errors = new();
+                    }
+
+                    errors.AddRange(result.Exceptions.InnerExceptions);
+                }
+            }
+
+            GetVulnerabilityInfoResult? final =
+                knownVulnerabilities != null || errors != null
+                ? new(knownVulnerabilities, errors != null ? new AggregateException(errors) : null)
+                : null;
+            return final;
+        }
+
+        private int ParseAuditLevel()
+        {
+            string? auditLevel = _restoreAuditProperties.AuditLevel;
```

Comment:
Do we need to `.Trim()` the value?

Reply:
I'd almost rather that happen in the `setter` for `AuditLevel` if so, that way all the `.Equals` logic/etc gets the benefit.

Reply:
RestoreAuditProperties comes from the PackageSpec, which is the input parameter for the entire restore process. In other words, if we validate and throw exceptions when settings `RestoreAuditProperties.AuditLevel`, then no packages will be restored, no vulnerabilities will be checked. If there is anything else that would output warnings or errors, they won't be output until the customer fixes their project file first and then runs again.  I don't think that's a good experience for a dev tool (the compiler doesn't stop at the first syntax error), hence the PackageSpec data structure is a faithful representation of project inputs, and validation happens here.

Reply:
> Trim() is sanitation, not validation. How would Trim() cause an exception to be thrown?

Oops, I misunderstood. My thoughts were elsewhere.  I'll have a look how other parts of our "create a PackageSpec instance" code handles trimming.